hpx/parallel/algorithms/for_each.hpp

```diff
@@ -45,11 +46,33 @@ namespace hpx { namespace parallel { HPX_INLINE_NAMESPACE(v1)
             typename hpx::util::decay<F>::type& f_;
             typename hpx::util::decay<Proj>::type& proj_;
 
+            template <typename T>
+            HPX_HOST_DEVICE HPX_FORCEINLINE
+            typename std::enable_if<
+               !hpx::traits::is_value_proxy<T>::value
+            >::type
+            call(T && t)
+            {
+                hpx::util::invoke(f_,
+                    hpx::util::invoke(proj_, std::forward<T>(t)));
+            }
+
+            template <typename T>
+            HPX_HOST_DEVICE HPX_FORCEINLINE
+            typename std::enable_if<
+                hpx::traits::is_value_proxy<T>::value
+            >::type
+            call(T && t)
+            {
+                auto tmp = hpx::util::invoke(proj_, std::forward<T>(t));
```

Comment:
I looked once more at this code. The idea behind this is that the user supplied lambda should be called by passing the value proxy itself, not the value wrapped by it. The code as it is in this PR allows for both, a user supplied lambda which takes its argument by value and by (non-const) reference. Obviously, the lambda has to be generic (or a PFO).

In conclusion, I don't think we should apply the changes suggested by @K-ballo.